pysoa/server/action/switched.py

```diff
@@ -0,0 +1,126 @@
+
+        # noinspection PyUnresolvedReferences
+        if bases[0] is not object and (
+            not cls.switch_to_action_map or
+            not hasattr(cls.switch_to_action_map, '__iter__') or
+            _len(cls.switch_to_action_map) < 2 or
+            any(
+                True for i in cls.switch_to_action_map
+                if not hasattr(i, '__getitem__') or _len(i) != 2 or not is_switch(i[0]) or not callable(i[1])
+            )
+        ):
+            raise ValueError(
+                'Class attribute switch_to_action_map must be an iterable of at least two indexable items, each '
+                'with exactly two indexes, where the first element is a switch and the second element is an action '
+                '(callable).'
+            )
+
+        return cls
+
+
+@six.add_metaclass(SwitchedActionMetaClass)
+class SwitchedAction(object):
+    """
+    A specialized action that defers to other, concrete actions based on request switches. Subclasses must not
+    override any methods and must override `switch_to_action_map`. `switch_to_action_map` should be some iterable
+    object that provides `__len__` (such as a tuple [recommended] or list). Its items must be indexable objects that
+    provide `__len__` (such as a tuple [recommended] or list) and have exactly two elements.
+
+    For each item in `switch_to_action_map`, the first element must be a switch that provides `__int__` (such as an
+    actual integer) or a switch that provides an attribute `value` which, itself, provides `__int__` (or is an int).
+    The second element must be an action, such as an action class (e.g. one that extends `Action`) or any callable
+    that accepts a server settings object and returns a new callable that, itself, accepts an `ActionRequest` object
+    and returns an `ActionResponse` object or raises an `ActionError`.
+
+    `switch_to_action_map` must have at least two items in it. `SwitchedAction` will iterate over that list, checking
+    the first element (switch) of each item to see if it is enabled in the request. If it is, the second element (the
+    action) of that item will be deferred to. If it finds no items whose switches are enabled, it will use the very
+    last action in `switch_to_action_map`. As such, you can treat the last item as a default, and its switch could
+    simply be `SwitchedAction.DEFAULT_ACTION` (although, this is not required: it could also be a valid switch, and
+    it would still be treated as the default in the case that no other items matched).
+
+    Example usage:
+
+    .. code-block:: python
+
+        class UserActionV1(Action):
+            ...
+
+        class UserActionV2(Action):
+            ...
+
+        class UserTransitionAction(SwitchedAction):
+            switch_to_action_map = (
+                (USER_VERSION_2_ENABLED, UserActionV2),
+                (SwitchedAction.DEFAULT_ACTION, UserActionV1),
+            )
+    """
+
+    DEFAULT_ACTION = -1827739192
```

Comment:
any reason why not just `-1`, given that switches are positive integers?

Reply:
_Our_ switches are positive integers. There's no hard requirement in SOA that _all_ switches be positive. The only requirement is "integer."

Reply:
OK, so the idea is that it's just a random number, that by chance will not be used?
Then I vote for `-802549514462` ;-)

Reply:
Is there a reason this has to be an `int`, rather than something obviously unique like a brand new class (what `NOT_PROVIDED` is powered by in Django) or `None`?

Reply:
It can't be as trivial as this unfortunately 
```In [1]: class Foo:
   ...:     def __int__(self):
   ...:         return None
   ...:

In [2]: foo = Foo()

In [3]: int(foo)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-3bf93630f200> in <module>()
----> 1 int(foo)

TypeError: __int__ returned non-int (type NoneType)```

Reply:
This will work though:

```
class Foo(int):
    def __eq__(self, other):
        return False
```

Reply:
Alright. Attempt 3. How bow now?